shipgen/ALPACAGenerator.cxx

```diff
@@ -0,0 +1,88 @@
+#include "ALPACAGenerator.h"
+#include "TDatabasePDG.h"               // for TDatabasePDG
+#include "TMath.h"                      // for Sqrt
+
+const Double_t cm = 10.; // pythia units are mm
+const Double_t c_light = 2.99792458e+10; // speed of light in cm/sec (c_light   = 2.99792458e+8 * m/s)
+
+using std::cout;
+using std::endl;
+// read events from ntuples produced
+
+// -----   Default constructor   -------------------------------------------
+ALPACAGenerator::ALPACAGenerator() {}
+// -------------------------------------------------------------------------
+// -----   Default constructor   -------------------------------------------
+Bool_t ALPACAGenerator::Init(const char* fileName) {
+  return Init(fileName, 0);
+}
+// -----   Default constructor   -------------------------------------------
+Bool_t ALPACAGenerator::Init(const char* fileName, const int firstEvent) {
+  cout << "Info ALPACAGenerator: Opening input file " << fileName << endl;
+  fInputFile  = new TFile(fileName);
+  if (fInputFile->IsZombie()) {
+    cout << "-E ALPACAGenerator: Error opening the Signal file" << fileName << endl;
+  }
+  fTree = (TTree *)fInputFile->Get("MCTrack");
+  fNevents = fTree->GetEntries();
+  fn=firstEvent;
+  fTree->SetBranchAddress("event",&event);
+  fTree->SetBranchAddress("track",&track);
+  fTree->SetBranchAddress("pdg",&pdg);              // particle pdg
+  fTree->SetBranchAddress("parent",&parent);    // parent track number??
+  fTree->SetBranchAddress("tof",&tof); // time of flight
+  fTree->SetBranchAddress("e",&e);          // incoming muon energy
+  fTree->SetBranchAddress("w",&w);                  // weight of event
+  fTree->SetBranchAddress("x",&x); // position
+  fTree->SetBranchAddress("y",&y);
+  fTree->SetBranchAddress("z",&z);
+  fTree->SetBranchAddress("px",&px);     // momentum
+  fTree->SetBranchAddress("py",&py);
+  fTree->SetBranchAddress("pz",&pz);
+  return kTRUE;
+}
+// -------------------------------------------------------------------------
+
+
+// -----   Destructor   ----------------------------------------------------
+ALPACAGenerator::~ALPACAGenerator()
+{
+ // cout << "destroy Ntuple" << endl;
+ fInputFile->Close();
+ fInputFile->Delete();
+ delete fInputFile;
+}
+// -------------------------------------------------------------------------
+
+// -----   Passing the event   ---------------------------------------------
+Bool_t ALPACAGenerator::ReadEvent(FairPrimaryGenerator* cpg)
+{
+  while (fn<fNevents) {
+  fTree->GetEntry(fn);
+  if (track==0&&parent==-1) {cpg->AddTrack(pdg,px,py,pz,x,y,z,-1.,false,e,tof,w);} //Be sure the order, units, and possible errors(i.e pdg/track_id)
+  fTree->GetEntry(fn+1);
+  if (track==1&&parent==0) {cpg->AddTrack(pdg,px,py,pz,x,y,z,0,true,e,tof,w);}
+  fTree->GetEntry(fn+2);
+  if (track==2&&parent==0) {cpg->AddTrack(pdg,px,py,pz,x,y,z,0,true,e,tof,w);}
+  fn=fn+3;
+  break;
+  }
+  //if (fn==fNevents) {
+     //cout << "No more input events"<<endl;
+     //return kFALSE; }
```

Comment:
This would still be good to fix, as it will currently crash in a bad way if the file is shorter than the number of events requested...

Reply:
have you had a look at why this doesn't work anymore? I guess `fNevents` could be skipped over if it's not a multiple of 3. So maybe this can be changed to `fn>=fNevents`?